File: trie/trie.go
package main

import "fmt"

type TrieNode struct {
    children  map[rune]*TrieNode
    isEndWord bool
}

type Trie struct {
    root *TrieNode
}

func NewTrie() *Trie {
    return &Trie{
        root: &TrieNode{
            children:  make(map[rune]*TrieNode),
            isEndWord: false,
        },
    }
}

func (t *Trie) Insert(word string) {
    node := t.root
    for _, char := range word {
        if _, exists := node.children[char]; !exists {
            node.children[char] = &TrieNode{
                children:  make(map[rune]*TrieNode),
                isEndWord: false,
            }
        }
        node = node.children[char]
    }
    node.isEndWord = true
}

func (t *Trie) Search(word string) bool {
    node := t.root
    for _, char := range word {
        if _, exists := node.children[char]; !exists {
            return false
        }
        node = node.children[char]
    }
    return node.isEndWord
}

func (t *Trie) StartsWith(prefix string) bool {
    node := t.root
    for _, char := range prefix {
        if _, exists := node.children[char]; !exists {
            return false
        }
        node = node.children[char]
    }
    return true
}

func main() {
    trie := NewTrie()
    words := []string{"apple", "app", "application", "apply", "banana"}
    
    for _, word := range words {
        trie.Insert(word)
    }
    
    fmt.Printf("Search 'app': %t\n", trie.Search("app"))
    fmt.Printf("Starts with 'app': %t\n", trie.StartsWith("app"))
}

File: trie/trie.cpp
#include <iostream>
#include <unordered_map>
#include <vector>
#include <string>
using namespace std;

class TrieNode {
public:
    unordered_map<char, TrieNode*> children;
    bool isEndWord;
    
    TrieNode() : isEndWord(false) {}
};

class Trie {
private:
    TrieNode* root;
    
public:
    Trie() {
        root = new TrieNode();
    }
    
    void insert(const string& word) {
        TrieNode* node = root;
        for (char c : word) {
            if (node->children.find(c) == node->children.end()) {
                node->children[c] = new TrieNode();
            }
            node = node->children[c];
        }
        node->isEndWord = true;
    }
    
    bool search(const string& word) {
        TrieNode* node = root;
        for (char c : word) {
            if (node->children.find(c) == node->children.end()) {
                return false;
            }
            node = node->children[c];
        }
        return node->isEndWord;
    }
    
    bool startsWith(const string& prefix) {
        TrieNode* node = root;
        for (char c : prefix) {
            if (node->children.find(c) == node->children.end()) {
                return false;
            }
            node = node->children[c];
        }
        return true;
    }
};

int main() {
    Trie trie;
    vector<string> words = {"apple", "app", "application", "apply", "banana"};
    
    for (const string& word : words) {
        trie.insert(word);
    }
    
    cout << "Search 'app': " << trie.search("app") << endl;
    cout << "Starts with 'app': " << trie.startsWith("app") << endl;
    
    return 0;
}
